edb/server/compiler/sertypes.py

```diff
@@ -65,11 +66,36 @@
 CTYPE_ANNO_TYPENAME = b'\xff'
 
 
+class Cardinality(enum.Enum):
+    AT_MOST_ONE = 0x0
+    ONE = 0x1
+    MANY = 0x2
+    AT_LEAST_ONE = 0x3
+
+    @classmethod
+    def from_ptr(cls, ptr, schema) -> Cardinality:
+        required = ptr.get_required(schema)
+        is_multi = ptr.get_cardinality(schema).is_multi()
+
+        if not required and not is_multi:
+            return cls.AT_MOST_ONE
+        if required and not is_multi:
+            return cls.ONE
+        if not required and is_multi:
+            return cls.MANY
+        if required and is_multi:
+            return cls.AT_LEAST_ONE
+
+        raise RuntimeError("unreachable")
+
+
 class TypeSerializer:
 
     EDGE_POINTER_IS_IMPLICIT = 1 << 0
     EDGE_POINTER_IS_LINKPROP = 1 << 1
     EDGE_POINTER_IS_LINK = 1 << 2
+    EDGE_POINTER_IS_REQUIRED = 1 << 3
```

Comment:
This seems to be dead code.

Reply:
:facepalm:

Reply:
https://github.com/edgedb/edgedb/pull/2612